### ISSUE TITLE
TextField onChange called with value not formatted by TextInputFormatter

### ISSUE THREAD

**Issue by spkersten** [flutter/flutter#58166]
I have a material `TextField` with this input formatter (prints added for debugging):

```dart
class Utf8BytesLimitingTextInputFormatter extends TextInputFormatter {
  Utf8BytesLimitingTextInputFormatter(this.maxBytes) : assert(maxBytes == null || maxBytes == -1 || maxBytes > 0);

  final int maxBytes;

  @override
  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
    print("oldValue: $oldValue");
    print("newValue: $newValue");
    print("utf8: ${utf8.encode(newValue.text)}  LEN: ${utf8.encode(newValue.text).length}" );
    if (maxBytes != null && maxBytes > 0 && utf8.encode(newValue.text).length > maxBytes) {
      print("Huh");
      return oldValue;
    }
    print("Ok");
    return newValue;
  }
}

...

TextField(
  inputFormatters: [
     Utf8BytesLimitingTextInputFormatter(32),
  ],
  onChange: (value) {
    print("Weird value: $value");
    widget.viewModel.changeName(value);
  },
)
```

Debug output:
```
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 26, extentOffset: 26, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 26))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 27, extentOffset: 27, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 27))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117]  LEN: 27
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvu
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 27, extentOffset: 27, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 27))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 28, extentOffset: 28, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 28))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118]  LEN: 28
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuv
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 28, extentOffset: 28, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 28))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 29, extentOffset: 29, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 29))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117]  LEN: 29
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvu
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 29, extentOffset: 29, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 29))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 30, extentOffset: 30, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 30))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118]  LEN: 30
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuv
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 30, extentOffset: 30, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 30))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 31, extentOffset: 31, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 31))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117]  LEN: 31
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvu
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 31, extentOffset: 31, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 31))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 32, extentOffset: 32, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 32))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118]  LEN: 32
I/flutter (13806): Ok
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvuv
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 32, extentOffset: 32, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 32))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvuvu├, selection: TextSelection(baseOffset: 33, extentOffset: 33, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 11, end: 33))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117]  LEN: 33
I/flutter (13806): Huh
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvuv
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvuvu
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@bafc3d4 mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=67
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@ee8f072 mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=60
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@e2f4940 mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=68
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper(13806): requestCursorAnchorInfo on inactive InputConnection
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper(13806): requestCursorAnchorInfo on inactive InputConnection
W/IInputConnectionWrapper(13806): setComposingText on inactive InputConnection
I/flutter (13806): oldValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvuv├, selection: TextSelection(baseOffset: 32, extentOffset: 32, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I/flutter (13806): newValue: TextEditingValue(text: ┤Test guvug vuvuvuvuvuvuvuvuvuvuvU├, selection: TextSelection(baseOffset: 33, extentOffset: 33, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 32, end: 33))
I/flutter (13806): utf8: [84, 101, 115, 116, 32, 103, 117, 118, 117, 103, 32, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 117, 118, 85]  LEN: 33
I/flutter (13806): Huh
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvuv
I/flutter (13806): Weird value: Test guvug vuvuvuvuvuvuvuvuvuvuvU
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@1d1a1be mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=60
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@45fb96c mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=67
D/InputMethodManager(13806): prepareNavigationBarInfo() DecorView@a73807[MainActivity]
D/InputMethodManager(13806): getNavigationBarColor() -16777216
V/InputMethodManager(13806): Starting input: tba=com.company.product.beta ic=io.flutter.plugin.editing.InputConnectionAdaptor@3021bca mNaviBarColor -16777216 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : false
D/InputMethodManager(13806): startInputInner - Id : 0
I/InputMethodManager(13806): startInputInner - mService.startInputOrWindowGainedFocus
D/InputTransport(13806): Input channel destroyed: 'ClientS', fd=60
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper(13806): requestCursorAnchorInfo on inactive InputConnection
W/IInputConnectionWrapper(13806): setComposingText on inactive InputConnection
W/IInputConnectionWrapper(13806): getExtractedText on inactive InputConnection
W/IInputConnectionWrapper(13806): requestCursorAnchorInfo on inactive InputConnection
W/IInputConnectionWrapper(13806): setComposingText on inactive InputConnection
```

When the text is at the maximum, the formatter is called and returns the `oldValue`. Next `onChange` is once called "correctly" (well, the value is the same as before: #47191, #44979) with the old value, but then another time incorrectly with the new value. The additional character is not shown on screen (which is as expected).

Also note, that the second time a character is typed while the text is its max, the last character in newValue is a capital `U`, while a lower case one was typed.

This was observed on a Samsung S9+(OS-10) and can be reproduced consistently. It could not be reproduced on either iPhone, iPhone Simulator or Android Simulator.

```
[✓] Flutter (Channel unknown, v1.17.1, on Mac OS X 10.15.4 19E287, locale en-NL)
    • Flutter version 1.17.1 at ...
    • Framework revision f7a6a7906b (2 weeks ago), 2020-05-12 18:39:00 -0700
    • Engine revision 6bc433c6b6
    • Dart version 2.8.2
```

**Comment by VladyslavBondarenko** [comment]
@spkersten 
It doesn't reproduce with for example my Redmi Note 7.
But @TahaTesser said that he reproduced the double calling `onChange` with Samsung Galaxy M30 running with current stable `1.17.1`. And he couldn't reproduce this when running with current dev `1.19.0-1.0.pre`. 
It may be fixed. Could you check with your device?

**Comment by pedromassango** [comment]
I noticed this while working this PR https://github.com/flutter/flutter/pull/57264

@justinmc can you confirm if I can try to fix this in the same PR or send nother one?

**Comment by pedromassango** [comment]
> And he couldn't reproduce this when running with current dev `1.19.0-1.0.pre`.

Tested on the current master, and I can confirm that this was fixed (I checked the code)

**Comment by justinmc** [comment]
Yup seems fixed to me, I'll go ahead and close this.

**Comment by lifenautjoe** [comment]
Is this already on stable? Is there any kind of temporal solution for this?

**Comment by justinmc** [comment]
I believe this should have been released with yesterday's v1.20.0 stable release.  I don't know what change exactly fixed this issue, but given the dates in this issue it should be in.
